### PR TITLE
Features/support functions with hook #179

### LINKED ISSUE #179: Support some functions with hooks
**Is your feature request related to a problem? Please describe.**
Should provide an ability to run some functions before all step, before each step, after all step and after each step, e.g:
```
afterAll(fn, timeout)
afterEach(fn, timeout)
beforeAll(fn, timeout)
beforeEach(fn, timeout)
```

### PR DESCRIPTION
**Pull request by ai-flood**
resolve #179

### CODE REVIEW

File: examples/hook/test.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		await browser.wait(Until.elementTextMatches('span.token', /\d+/))
+		let element = await browser.findElement('span.token')
+		let token = await element.text()
+		await browser.type(By.id('challenger_one_time_token'), token)
+
+		await browser.takeScreenshot()
+		await browser.click('input.btn')
+	})
+
+	step('Flood Challenge: Step 5', async (browser: Driver) => {
```
Comment (HieuLeKMS) [inline examples/hook/test.ts:82]: I think we should run just 2 or 3 steps, that's enough to see how the hooks work

File: packages/core/src/runtime/Hook.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	beforeAll: HookBase[]
+	beforeEach: HookBase[]
+}
+
+/**
+ * @internal
+ */
+export function normalizeHookBase(hookBase: HookBase): HookBase {
+	const { waitTimeout } = hookBase
+	hookBase.waitTimeout = waitTimeout && waitTimeout > 0 ? waitTimeout : 30 * 1e3
```
Comment (HieuLeKMS) [inline packages/core/src/runtime/Hook.ts:26]:
```suggestion
	waitTimeout = waitTimeout && waitTimeout > 0 ? waitTimeout : 30 * 1e3
```

File: packages/core/src/runtime/Test.ts [hunk 3]
```diff
@@ -396,4 +410,33 @@ export default class Test implements ITest {
...
+
+	private async runHookFnc(
+		hooks: HookBase[],
+		browser: Browser<Step>,
+		testDataRecord: any,
+	): Promise<void> {
+		try {
+			for (const hook of hooks) {
+				browser.settings = { ...this.settings }
+				browser.settings.waitTimeout = hook.waitTimeout
```
Comment (phileba) [inline packages/core/src/runtime/Test.ts:422]: Do we need to restore the raw `waitTimeout` after running those hook tests? I think the scope of this setting is just in the step hook, not in other kinds of steps

Comment (ai-flood) [reply packages/core/src/runtime/Test.ts:422]: Sure, the setting of step always assigned at line 324 in function runStep

Comment (phileba) [reply packages/core/src/runtime/Test.ts:422]: cool

File: packages/core/src/runtime/Test.ts [hunk 4]
```diff
@@ -414,4 +429,33 @@ export default class Test implements ITest {
...
+				const hookFunc = hook.fn.bind(null, browser, testDataRecord)
+				await this.doHookFncWithTimeout(hookFunc, hook.waitTimeout)
+			}
+		} catch (error) {
+			throw new Error(error)
+		}
+	}
+
+	private async doHookFncWithTimeout(func: any, timeout: number): Promise<any> {
```
Comment (ivanvanderbyl) [inline packages/core/src/runtime/Test.ts:450]: Please use consistent naming for functions. We try not to abbreviate names, except for function = fn

Comment (ai-flood) [reply packages/core/src/runtime/Test.ts]: Resolved

File: packages/core/src/runtime/Hook.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { Browser } from './types'
```
Comment (ivanvanderbyl) [inline packages/core/src/runtime/Hook.ts:1]: I think we should rename this package `lifecycle`, hook is an overloaded term these days.

**Pull request merged**